archaius2-core/src/main/java/com/netflix/archaius/DefaultPropertyFactory.java

```diff
@@ -321,46 +326,48 @@ public KeyAndType(String key, Type type) {
         }
 
         public <S> KeyAndType<S> discardType() {
-            return new KeyAndType<S>(key, null);
+            if (type == null) {
+                @SuppressWarnings("unchecked") // safe since type is null
+                KeyAndType<S> keyAndType = (KeyAndType<S>) this;
+                return keyAndType;
+            }
+            return new KeyAndType<>(key, null);
         }
 
         public KeyAndType<T> withKey(String newKey) {
-            return new KeyAndType<T>(newKey, type);
+            return new KeyAndType<>(newKey, type);
         }
-        
+
         public boolean hasType() {
             return type != null;
         }
 
         @Override
         public int hashCode() {
-            final int prime = 31;
             int result = 1;
-            result = prime * result + ((key == null) ? 0 : key.hashCode());
-            result = prime * result + ((type == null) ? 0 : type.hashCode());
+            result = 31 * result + Objects.hashCode(key);
+            result = 31 * result + Objects.hashCode(type);
             return result;
```

Comment:
Couldn't we just return `Objects.hash(key, type)`?

Reply:
I didn't use it because it's a vararg method, so allocates an array. Maybe it doesn't matter here, or would get inlined by the JIT anyway, I don't know.